### PR TITLE
Improve creating registry keys from observatory names listed in JSON

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

The file https://github.com/astropy/astropy-data/blob/84deb2cda3a62a852950fc766b9b16fd11296de3/coordinates/sites.json used by the machinery in `astropy/coordinates/sites.py` contains `name` and `aliases` entries for each site, but `astropy` currently does not use the `name` entry as a registry key but only as metadata. Many entries in the JSON file therefore duplicate the preferred name of the observatory both in `name` and `aliases`, and the ones that do not cannot be accessed by the preferred name through `astropy.coordinates.EarthLocation.of_site()`. However, `astropy` does treat the empty string as a valid alias for the observatory name.

This pull request improves creating registry keys from observatory names, and it also makes it possible to remove duplicated `aliases` entries in the future.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Might need a change log?

File: astropy/coordinates/sites.py
Comment (hamogu) [inline astropy/coordinates/sites.py:112]: I wonder if it makes sense to raise an error or a warning here instead of silently dropping names? I know that this was no treated properly before, so while making it better, let's make is as good as we can!
I see two possible problems that may or may not turn up in the current sites.json file:
- An observatory is added twice and either name or aliases overlap.
- Two different observatories register the same alias or name. While that should not happen for big observatories (there is only one VLT) or long names ("Tautenburger Sternwarte"), I would not be surprised if the same three-letter acronym is used for different observatories in different parts of the world.

Either condition should not happen in a well curated sites.json, but (a) mistakes happen, (b) the user could supply their own sites.json and (c) it's very easy to just just an "else: raise XXX" in here. 

I suggest that a warning is sufficient, since sites.json is long and an error in observatory "SWO" should not stop me from getting coordinates for a totally unlates query like "Gemini/North".

Comment (eerovaher) [reply astropy/coordinates/sites.py:112]: Raising warnings about duplicated names sounds like a good idea, but it should not be implemented until the `sites.json` file in `astropy-data` is cleaned from duplicates, otherwise the users would receive many warnings that they can't really do anything about. Removing the duplicates from `sites.json` would reduce the number of usable keys for anyone not using the bleeding edge version of `astropy`, so it doesn't seem it would be a good idea to do that quite yet.

When it comes to checking for duplicated names for different sites then it wouldn't be too difficult to add a test to `astropy-data` that would check for that. The output of `astropy.coordinates.EarthLocation.get_site_names()` shows that currently there aren't any duplicated labels across sites other than the empty string that this pull request takes care of.

**Comment by eerovaher** [comment]
I noticed that there are a few instances where the `name` attribute of a site entry matches its key in the `sites.json` file. I updated the code so that this will not cause duplicated keys in the registry.

**Comment by eerovaher** [comment]
I have extended this pull request so that there is now a tested example of using `EarthLocation.get_site_names()` where the function is mentioned in `docs/coordinates.index.rst`. I rebased the previous commits to incorporate the changes in `main` that ensure the documentation build succeeds also with `numpydoc` 1.2.

**Comment by pllim** [comment]
@eerovaher , still needs a change log.

And if we cannot get approval from @hamogu or @adrn by feature freeze, we should move the milestone. Thanks!

**Comment by hamogu** [comment]
@eerovaher I think this would be useful to have and I'm happy to review, approve and merge if CI passes,  if you put in the changelog first.

**Comment by adrn** [comment]
Agreed - sorry for the delay, but with a changelog entry I think this is ready to go in!

File: astropy/coordinates/sites.py
Comment (adrn) [reply astropy/coordinates/sites.py:112]: Could you open an issue over in astropy/astropy-data to remind about this?

Comment (pllim) [reply astropy/coordinates/sites.py:112]: There is a change log now (thanks!) and I opened issue at astropy/astropy-data#126 . Please do a final review of this PR and approve if satisfactory. Thanks!

File: docs/coordinates/index.rst
Comment (adrn) [inline docs/coordinates/index.rst:3]: I'm actually confused about this comment: it seems like in the existing documentation, `of_site()` isn't actually called below (that doctest on L339 is skipped?). But I think that might be ok: given the issues we have with remote data tests and the stability of the astropy-data server, we might consider keeping it as is and removing this setup instead so that this page doesn't actually make any additional remote data calls. Thoughts? (and cc @pllim)

Comment (pllim) [reply docs/coordinates/index.rst:3]: Is this setup not needed anywhere else in the remote data doctest suite?

I'd rather you use the local data file but I don't know the *intent* of this test.

Comment (pllim) [reply docs/coordinates/index.rst:3]: My vote is: Less remote data the better, if possible.

Comment (adrn) [reply docs/coordinates/index.rst:3]: > Is this setup not needed anywhere else in the remote data doctest suite?

It looks like most calls to `of_site()` in the docs are skipped, but there are a few that aren't, so we can address this aspect in a separate PR/issue.

File: docs/coordinates/index.rst
Comment (adrn) [inline docs/coordinates/index.rst:345]:
```suggestion
    >>> EarthLocation.get_site_names()  # doctest: +SKIP
```

Comment (eerovaher) [reply docs/coordinates/index.rst:345]: I uncovered the problem this pull request is meant to solve by implementing this test, so it clearly serves a purpose and I'd rather not skip it.

Comment (eteq) [reply docs/coordinates/index.rst:345]: I agree with @eerovaher here!

File: docs/coordinates/index.rst
Comment (eteq) [reply docs/coordinates/index.rst:3]: I agree with @eerovaher's point that this caught a real genuine bug, so while I agree with "less remote data the better, if possible", I think in this case it's really "minimal necessary to catch bugs", and this is catching bugs!

**Review by eteq — APPROVED** [review]
I like this as it stands now. I'll try to address astropy/astropy-data#126 between now and release if I can, since I personally think it's legit to consider fixing that as a bug. But even if not, this is still an improvement as it stands.

I also read @adrn's https://github.com/astropy/astropy/pull/12721#issuecomment-1105895141 as effectively an approval (although I'll also approve), so I'm going to merge this now to start clearing the backlog.

**Pull request merged**
